enrichments/trace/internal/elastic/span.go

```diff
@@ -0,0 +1,294 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package elastic
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	semconv "go.opentelemetry.io/collector/semconv/v1.25.0"
+	"google.golang.org/grpc/codes"
+)
+
+// Span is a representation of OTel `ptrace.Span`. The struct is used
+// to enrich a span with Elastic specific attributes. These attributes
+// are derived from the base attributes and appended to the span
+// attributes. The enrichment logic is performed by categorizing the
+// OTel spans into 2 different categories:
+//   - Elastic transactions, defined as spans which measure the highest
+//     level of work being performed with a service.
+//   - Elastic spans, defined as all spans (including transactions).
+//     However, for the enrichment logic spans are treated as a separate
+//     entity i.e. all transactions are not enriched as spans and vice versa.
+type Span struct {
+	urlFull *url.URL
+
+	peerService     string
+	urlScheme       string
+	urlDomain       string
+	urlPath         string
+	urlQuery        string
+	rpcSystem       string
+	rpcService      string
+	grpcStatus      string
+	dbName          string
+	dbType          string
+	messagingSystem string
+	messagingDestn  string
+
+	urlPort        int64
+	httpStatusCode int64
+
+	spanStatusCode ptrace.StatusCode
+
+	isMessaging        bool
+	isRPC              bool
+	isHTTP             bool
+	isDB               bool
+	messagingDestnTemp bool
+}
+
+func (s *Span) Enrich(span ptrace.Span) {
```

Comment:
Given that this is the only accessible part of the API, I think it may be worth changing the API a bit to be like:

```go
func EnrichSpan(span ptrace.Span) {
	var c spanEnrichmentContext
	c.enrich(span)
}

type spanEnrichmentContext {
	urlFull *url.URL

	peerService     string
	...
}

func (c *spanEnrichmentContext) enrich(span ptrace.Span) {
	...
}
```